### ISSUE TITLE
Scrolling down not working the same on Brave and Firefox

### ISSUE THREAD

**Issue by tommyalatalo** [gdh1995/vimium-c#592]
**Describe the bug**

I have vimium c installed on both Brave and Firefox, but scrolling down on the page (I have mapped the up and down arrows to this) doesn't smooth scroll properly on Brave, instead it just jumps one 'notch' down, and then stops. The exact same vimium config on Firefox works just fine, scrolling both up and down works continuously and smoothly.

**To Reproduce**

1. Install vimium c on brave 
2. 
```
map <up> scrollUp
map <down> scrollDown
```
3. The scrollDown doesn't run continuously but stops after the first increment.

While I can't entirely rule out that there isn't another setting somewhere interfering with this, I would still need some pointers for how to troubleshoot this. I think this is a behavior happening in Brave though.

**Browser, OS and Vimium C versions**

- Browser name: Brave
- Browser version: Version 1.36.122 Chromium: 99.0.4844.88 (Official Build) (64-bit)
- Vimium C version: 1.97.0
- OS name and version: Arch Linux

**Comment by gdh1995** [comment]
So, did you mean "on brave Vimium C can not scroll by expected distance (pixels)"?

**Comment by tommyalatalo** [comment]
No, I mean that if I hold the down arrow mapped for scrollDown, then on Firefox it continues scrolling continuously until I release the key (as expected), but in Brave it just scrolls one increment and then immediately stops.

**Comment by gdh1995** [comment]
Does the default `j` and `k` scroll successfully when you hold one of them down?

It sounds like Vimium C didn't receive the message that a `down` key was repeated. If my guess is correct, then it's a compatibility bug of Brave on your version of Linux .

**Comment by tommyalatalo** [comment]
Holding `J` and `K` behave the same way as the arrow keys do; scrolling down with `J` stops after one increment while scrolling up wit `K` works correctly and keeps scrolling until the key is released.

I just tested it on another computer with the same OS and Brave, on that host scrolling both up and down work fine, so it's not a bug in Brave it would seem.

**Comment by gdh1995** [comment]
Have no idea. Please test this:
1. enable smooth scrolling
2. add ```map j scrollDown``` and ```map k scrollUp`
3. open https://cn.bing.com/search?q=test&ensearch=1&FORM=BESBTB
4. click a blank area such as the middle-right side
5. press `j` and hold on to see how Vimium C scrolls

**Comment by tommyalatalo** [comment]
Works perfectly on Firefox, doesn't work properly according to my previous description description on Brave. This is on one computer, when testing on another it does work as expected.

> Have no idea. Please test this:
> 
>     1. enable smooth scrolling
> 
>     2. add `map j scrollDown` and `map k scrollUp` on Vimium C Options
> 
>     3. open https://cn.bing.com/search?q=test&ensearch=1&FORM=BESBTB
> 
>     4. click a blank area such as the middle-right side
> 
>     5. press `j` and hold on to see how Vimium C scrolls

Works perfectly on Firefox, doesn't work properly according to my previous description on Brave. This is on one computer, when testing on another it does work as expected on Brave as well.

**Comment by gdh1995** [comment]
Not reproduced on Brave 1.37 (Chromium 100) on Ubuntu 20.04.

The most possible reason is another process on your system hooks such long-pressed keys so Vimium C couldn't receive those keyboard event, or another browser extension did so.

To check whether Vimium C receives the info or not, please:
1. open https://gdh1995.cn/vimium-c/keyboard-test.html on Brave
2. press tons of `ArrowDown` to make the page very long (high)
3. resize your browser window and make it small, so the page shows a scrollbar
4. press down `j` and hold on, see what changes the page will show

**Comment by Enayaaa** [comment]
I am using Vivaldi, I have the same issue of holding down `j` doesn't keep scrolling, but only jumps a little bit and then stops. `k` works fine and continues scrolling up if I hold down.

I have tried the following:
> I Have no idea. Please test this:
>
> 1. enable smooth scrolling
> 2. add map j scrollDown and map k scrollUp on Vimium C Options
> 3. open https://cn.bing.com/search?q=test&ensearch=1&FORM=BESBTB
> 4. click a blank area such as the middle-right side press j and hold on to see how Vimium C scrolls

What you have described works completely fine on your website and a website I made real quick with alot of text and open.kattis.com, BUT it doesn't work on github.com for example, youtube.com, google.com and so on.

So the problem seems to be with certain websites. I can't figure out what it could be caused by.

**Comment by tommyalatalo** [comment]
I've also noticed that it seems to be dependant on which site sometimes, though strangely enough it now works on Brave as well, on the page github.com//issues/592 too which @Enayaaa mentioned. 

I haven't changed any settings since last posting here, so there must have been an update pushed somewhere.

**Comment by tommyalatalo** [comment]
No, it broke again on Github, doesn't scroll continuously again.
[Version 1.37.113 Chromium: 100.0.4896.88 (Official Build) (64-bit)](https://brave.com/latest/)

**Comment by gdh1995** [comment]
Sorry but I just find I didn't get the error. When you said "scroll by **one increment** and stop", what're the details?

Let's say about triggering `scrollPageDown` (or `scrollFullPageDown`) by `map j scrollPageDown` and `j`. Here's what it should do:
1. it should scroll down by half of a screen height - if the screen height is 1080px, then it should scroll by 540px
2. when you hold on the key, Vimium C should keep scrolling until it reaches the bottom
3. if the scrolling target is only a page element, but you're still holding on the key mapped to `scroll*`, then
- 1. Vimium C should try to find an ancestor (upper level) of the element to scroll
- 2. if there's no such element, then make a whole page scroll
4. and you may click some page elements to notify Vimium C what should scroll - a blank area usually means a whole page

Then when this bug happens (firstly please ensure Vimium C is trying to make a whole page scroll),
1. Will a single `j` (by pressing down and up at once) scroll by 540px? Or only tens of pixels ?
2. What will it do if you keep `j` holding on for seconds?
- 1. in my imagination, it may scroll down by second step in timestamp=1.0 second, and then a third in timestamp=2.0 second
  - 1. this means Vimium C receives a new keydown event of `j` and trigger `scrollPageDown` again
- 2. or it may just do nothing - looks like it thinks itself is still make something scroll
  - 1. this means Vimium C is trapped in a "scrolling" state

There're some other ways to trigger Vimium C's smooth scrolling, will these work?
1. when `LinkHints` shows, `ArrowDown` and `ArrowUp`
2. after `enterVisualMode` and in Visual Mode, `Ctrl+ArrowDown` and `Ctrl+ArrowUp`
3. after `enterFindMode` and when the Find HUD is focused, `Ctrl+ArrowDown` and `Ctrl+ArrowUp`
4. when `Vomnibar` is focused, `Ctrl+ArrowDown` and `Ctrl+ArrowUp`

Sorry for such a big list to test, but I seldom use Linux and can not reproduce this on my Ubuntu 20.04 in VMWare.

**Comment by NNiX1** [comment]
Hello just stumbled randomly on this thread and i have exactly same thing happening but only on one site that is [forum.rewasd.com](url) all other websites work normally but if you open any thread on rewasd forum then the exact thing happens when scrolling up everything is fine when trying to scroll down then it just jumps one and you need to be pressing J .. just though i post it so you can reproduce it reliably, IF its the same thing.

**Comment by NNiX1** [comment]
Hello just stumbled randomly on this thread and i have exactly same thing happening but only on one site that is [forum.rewasd.com](https://www.forum.rewasd.com) all other websites work normally but if you open any thread on rewasd forum then the exact thing happens when scrolling up everything is fine when trying to scroll down then it just jumps one and you need to be pressing J .. just though i post it so you can reproduce it reliably, IF its the same thing.
